app/lib/activitypub/activity/create.rb

```diff
@@ -165,40 +165,77 @@ def in_reply_to_uri
   end
 
   def text_from_content
+    return Formatter.instance.linkify([text_from_name, object_url].join(' ')) if converted_object_type?
+
     if @object['content'].present?
       @object['content']
-    elsif language_map?
+    elsif content_language_map?
       @object['contentMap'].values.first
     end
   end
 
-  def language_from_content
-    return LanguageDetector.instance.detect(text_from_content, @account) unless language_map?
-    @object['contentMap'].keys.first
+  def text_from_name
+    if @object['name'].present?
+      @object['name']
+    elsif name_language_map?
+      @object['nameMap'].values.first
+    end
+  end
+
+  def detected_language
+    if content_language_map?
+      @object['contentMap'].keys.first
+    elsif name_language_map?
+      @object['nameMap'].keys.first
+    elsif supported_object_type?
+      LanguageDetector.instance.detect(text_from_content, @account)
+    end
   end
 
   def object_url
     return if @object['url'].blank?
 
-    value = first_of_value(@object['url'])
+    # The url attribute can be a string, an array of strings, or an array of objects.
+    # The objects could include a mimeType. Not-included mimeType means it's text/html.
+    # We need to prefer text/html. So if the attribute has the fullest information,
+    # we've got to select an item that either has no mimeType or text/html mimeType
+    value = if @object['url'].is_a?(Array) && !@object['url'].first.is_a?(String)
```

Comment:
again, probably needs a function in JsonLdHelpers

Reply:
This is a more specific version of first_or_value, I can't easily generalize it.

Reply:
Isn't this the way we should process any link value though?